### ISSUE TITLE
"Cannot read property 'type' of undefined" after upgrading to 4.14.0

### ISSUE THREAD

**Issue by budarin** [eslint/eslint#9767]
Got the error after upgrading to 4.14.0 on Flow type definitions
```
TypeError: Cannot read property 'type' of undefined

TypeError: Cannot read property 'type' of undefined
    at isForInRef (..\node_modules\eslint\lib\rules\no-unused-vars.js:410:24)
    at variable.references.some.ref (..\node_modules\eslint\lib\rules\no-unused-vars.js:447:21)
    at Array.some (<anonymous>)
    at isUsedVariable (..\node_modules\eslint\lib\rules\no-unused-vars.js:446:40)
    at collectUnusedVariables (..\node_modules\eslint\lib\rules\no-unused-vars.js:569:26)
    at collectUnusedVariables (..\node_modules\eslint\lib\rules\no-unused-vars.js:576:17)
    at Program:exit (..\node_modules\eslint\lib\rules\no-unused-vars.js:621:36)
    at listeners.(anonymous function).forEach.listener (..\node_modules\eslint\lib\util\safe-emitter.js:47:58)
    at Array.forEach (<anonymous>)
    at Object.emit (..\node_modules\eslint\lib\util\safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (..\node_modules\eslint\lib\util\node-event-generator.js:251:26)
    at NodeEventGenerator.applySelectors (..\node_modules\eslint\lib\util\node-event-generator.js:280:22)
    at NodeEventGenerator.leaveNode (..\node_modules\eslint\lib\util\node-event-generator.js:303:14)
    at CodePathAnalyzer.leaveNode (..\node_modules\eslint\lib\code-path-analysis\code-path-analyzer.js:630:23)
    at CodePathAnalyzer.overrideLeaveNode (..\node_modules\eslint-plugin-node\lib\rules\process-exit-as-throw.js:127:27)
    at Traverser.leave [as _leave] (..\node_modules\eslint\lib\linter.js:1000:32)
Process finished with exit code -1
```

**Comment by platinumazure** [comment]
Are you using babel-eslint? If so, please try upgrading to the latest version of that parser.

**Comment by batazor** [comment]
I have the same problem

1. v4.14.0
2. https://github.com/micro-company/react-app/blob/f7763f4f47e636cc32e1ede2a523b665f50a82c3/.eslintrc.yaml
3.
```
> react-app@0.1.0 lint:fix /home/batazor/micro-company/react-app
> $(npm bin)/eslint --fix ./src

Cannot read property 'type' of undefined
TypeError: Cannot read property 'type' of undefined
    at isForInRef (/home/batazor/micro-company/react-app/node_modules/eslint/lib/rules/no-unused-vars.js:410:24)
    at variable.references.some.ref (/home/batazor/micro-company/react-app/node_modules/eslint/lib/rules/no-unused-vars.js:447:21)
    at Array.some (<anonymous>)
    at isUsedVariable (/home/batazor/micro-company/react-app/node_modules/eslint/lib/rules/no-unused-vars.js:446:40)
    at collectUnusedVariables (/home/batazor/micro-company/react-app/node_modules/eslint/lib/rules/no-unused-vars.js:569:26)
    at Program:exit (/home/batazor/micro-company/react-app/node_modules/eslint/lib/rules/no-unused-vars.js:621:36)
    at listeners.(anonymous function).forEach.listener (/home/batazor/micro-company/react-app/node_modules/eslint/lib/util/safe-emitter.js:47:58)
    at Array.forEach (<anonymous>)
    at Object.emit (/home/batazor/micro-company/react-app/node_modules/eslint/lib/util/safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (/home/batazor/micro-company/react-app/node_modules/eslint/lib/util/node-event-generator.js:251:26)
```
4. What you expected to happen instead: ` `

If you disable the `no-unused-vars` rule, everything works as expected

**Comment by platinumazure** [comment]
Seems like it might be a duplicate of #9762. Anyone think otherwise?

**Comment by batazor** [comment]
Yes, it looks like that.
However, the upgrade to `"babel-eslint ":" ^8.1.0 " did not solve the problem

**Comment by KevinGrandon** [comment]
I'm reproducing this with babel-eslint@8.1.0 and eslint@4.14.0.

**Comment by not-an-aardvark** [comment]
Could you provide an example of code which causes this issue?

**Comment by budarin** [comment]
I've installed "babel-eslint": "^8.1.0"

```js
// @flow

import * as React from 'react';

type Temp2Props = {
    count: number,
};

class Temp2 extends React.Component<Temp2Props> {
    shouldComponentUpdate(nextProps) {
        return this.props.count !== nextProps.count;
    }

    render() {
        return <div>{this.props.count}</div>;
    }
}

export default Temp2;
```

**Comment by ndelangen** [comment]
Can also confirm this is an issue, upgrading caused this:
```
> $ yarn lint                                                                          [±master ●]
yarn run v1.3.2
$ yarn lint:js . && yarn lint:md .
$ cross-env NODE_ENV=production eslint --cache --cache-location=.cache/eslint --ext .js,.jsx,.json .
Cannot read property 'type' of undefined
TypeError: Cannot read property 'type' of undefined
    at isForInRef (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/rules/no-unused-vars.js:410:24)
    at variable.references.some.ref (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/rules/no-unused-vars.js:447:21)
    at Array.some (<anonymous>)
    at isUsedVariable (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/rules/no-unused-vars.js:446:40)
    at collectUnusedVariables (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/rules/no-unused-vars.js:569:26)
    at collectUnusedVariables (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/rules/no-unused-vars.js:576:17)
    at Program:exit (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/rules/no-unused-vars.js:621:36)
    at listeners.(anonymous function).forEach.listener (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/util/safe-emitter.js:47:58)
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/util/safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/util/node-event-generator.js:251:26)
    at NodeEventGenerator.applySelectors (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/util/node-event-generator.js:280:22)
    at NodeEventGenerator.leaveNode (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/util/node-event-generator.js:303:14)
    at CodePathAnalyzer.leaveNode (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:630:23)
    at Traverser.leave [as _leave] (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/linter.js:1000:32)
    at Traverser._traverse (/Users/dev/Projects/GitHub/storybook/core/node_modules/eslint/lib/util/traverser.js:155:18)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by not-an-aardvark** [comment]
@budarin I can't seem to reproduce this issue with the code that you've provided. I'm using the following code in `foo.js` (the file you provided):

<details>
<summary>foo.js</summary>

```js
// @flow

import * as React from 'react';

type Temp2Props = {
    count: number,
};

class Temp2 extends React.Component<Temp2Props> {
    shouldComponentUpdate(nextProps) {
        return this.props.count !== nextProps.count;
    }

    render() {
        return <div>{this.props.count}</div>;
    }
}

export default Temp2;
```

</details>

With the following config file:

<details>
<summary>.eslintrc.yml</summary>

```yaml
parser: babel-eslint
rules:
  no-unused-vars: error
```

</details>

And I'm running the following from the command line:

```
$ rm -rf node_modules/
$ npm install eslint@4.14.0 babel-eslint@8.1.0
$ node_modules/.bin/eslint foo.js
```

When I do this, no errors are reported.

I don't doubt that some problem exists based on the number of people commenting on this issue, but I wasn't able to reproduce a problem with your example code.

**Comment by batazor** [comment]
@not-an-aardvark
My error occurs when I check the directory, but not when I check the files separately

**Comment by not-an-aardvark** [comment]
Are you sure that the file you pasted is the file where the crash is occurring? (To find the file where the crash is occurring, try running ESLint with the `--debug` flag to see which file is currently getting linted when the crash happens.)

**Comment by mysticatea** [comment]
I apology for the inconvenience. We have found a bug in `babel-eslint`'s new logic that I added. It does not work correctly if it falls back to monkeypatching once (If you are using a plugin which parses some files (e.g. imported files) with `babel-eslint`, it can fall back).

We have published `babel-eslint@8.1.1` to fix this. Please try it.

Further reading: babel/babel-eslint#558, babel/babel-eslint#559.

**Comment by budarin** [comment]
Thanks
It's solved the problem!
